### DISCUSSION TITLE
New options for Allow user signups - invite only

### DISCUSSION THREAD

**Discussion by GameBurrow** [Infisical/infisical#6314] [Feature Request]
### Feature description
Currently **Allow user signups** has 2 options:
- `Anyone`
- `Disabled`

We need third option - `Invite only`. Only allow registration for users that actually got invited.

### Why would it be useful?
Currently the only way to limit user registration is by domain only. That's useful for corporations/companies/businesses where all people use same emails, but for more casual use, where people cane have different email domains (for example I have a kube cluster that is shared by me and my close friends), I have to temporarely enable registrations, so I can invite those people, which I consider a security risk.  Especially considering you can't fully delete users through UI in Free tier, if some random person starts user registration process, while I have it open, it's annoying.

### Additional context
Strange thing is looking at old issue reports and forum posts, it seems like you had this feature using `INVITE_ONLY_SIGNUP` env flag but it got removed at one point? https://github.com/Infisical/infisical/issues/1892

**Comment by kasyap1234** [comment]
@maidul98 I would like to work on this issue .

**Comment by sheensantoscapadngan** [comment]
Hey @GameBurrow! When the **Allow user signups** option is Disabled, your users should still be able to register to the organization by clicking the invite link sent via email upon org invitation. At the moment, they will have to go through the invitation link for registration to work as expected. Let me know if you have any additional questions! 

<img width="497" alt="image" src="https://github.com/user-attachments/assets/39334fd3-0600-48cb-baec-b0b03528739b">

**Comment by GameBurrow** [comment]
@sheensantoscapadngan 

Hey. I tried it with 2 people. When they clicked the links, it just directed them to the login page. Both of them said that and tried their links myself, can confirm. Even tried to invitie myself with secondary email and same happens.

Link format was https://[redacted]/signupinvite?token=[redacted]&to=[redacted]&organization_id=[redacted]

**Comment by sheensantoscapadngan** [comment]
This is indeed a bug in the signup flow. I've raised a PR to address this [here](https://github.com/Infisical/infisical/pull/2466). 

Thank you for reporting this and for the prompt response!

**Comment by vwbusguy** [comment]
> Hey @GameBurrow! When the **Allow user signups** option is Disabled, your users should still be able to register to the organization by clicking the invite link sent via email upon org invitation. At the moment, they will have to go through the invitation link for registration to work as expected. Let me know if you have any additional questions!
> <img alt="image" width="497" src="https://private-user-images.githubusercontent.com/65645666/369685763-39334fd3-0600-48cb-baec-b0b03528739b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcyOTAxMzIsIm5iZiI6MTcyNzI4OTgzMiwicGF0aCI6Ii82NTY0NTY2Ni8zNjk2ODU3NjMtMzkzMzRmZDMtMDYwMC00OGNiLWJhZWMtYjBiMDM1Mjg3MzliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI1VDE4NDM1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNWNmMjM0ZjJlOWMzN2JiNzcxNTAzNDgwMTE5NGEzYTVkN2I4ODk5ZDFkYmM2MGQ5NGUzYzI5NTljZWY3OWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.D585nOMnsDii2HH1qgHPqP_8O2nLntl3YbbIxJFc100">

This caused problems for us today in onboarding a new user.  The invited user could not complete sign up until we enabled global sign up and then disabled it again once the user had completed sign in.  We also had INVITE_ONLY_STARTUP set in our environment, so this was an unexpected breaking change for us.  Currently (in v0.86.1), with Sign Up disabled, you can invite a user but that user will never be able to complete the sign in process, but will be stuck on the Sign Up page with a broken form.

**Comment by akhilmhdh** [comment]
Moving this to [feature request discussion](https://github.com/Infisical/infisical/issues/5172).

**Reply by GameBurrow** [reply]
@akhilmhdh If you actually go through the thread, you'll notice it was/is (haven't needed to test this as I used workaround to get around it) actually a bug not feature request.
